packages/@aws-cdk/aws-s3/lib/notifications-resource/notifications-resource-handler.ts

```diff
@@ -0,0 +1,165 @@
+ * required. We propagate errors and results as-is.
+ *
+ * Sadly, we can't use @aws-cdk/aws-lambda as it will introduce a dependency
+ * cycle, so this uses raw `cdk.Resource`s.
+ */
+export class NotificationsResourceHandler extends cdk.Construct {
+    /**
+     * Defines a stack-singleton lambda function with the logic for a CloudFormation custom
+     * resource that provisions bucket notification configuration for a bucket.
+     *
+     * @returns The ARN of the custom resource lambda function.
+     */
+    public static singleton(context: cdk.Construct) {
+        const root = cdk.Stack.find(context);
+
+        // well-known logical id to ensure stack singletonity
+        const logicalId = 'BucketNotificationsHandler050a0587b7544547bf325f094a3db834';
+        let lambda = root.tryFindChild(logicalId) as NotificationsResourceHandler;
+        if (!lambda) {
+            lambda = new NotificationsResourceHandler(root, logicalId);
+        }
+
+        return lambda.functionArn;
+    }
+
+    /**
+     * The ARN of the handler's lambda function. Used as a service token in the
+     * custom resource.
+     */
+    public readonly functionArn: cdk.Arn;
+
+    constructor(parent: cdk.Construct, id: string) {
+        super(parent, id);
+
+        const role = new iam.Role(this, 'Role', {
+            assumedBy: new cdk.ServicePrincipal('lambda.amazonaws.com'),
+            managedPolicyArns: [
+                cdk.Arn.fromComponents({
+                    service: 'iam',
+                    region: '', // no region for managed policy
+                    account: 'aws', // the account for a managed policy is 'aws'
+                    resource: 'policy',
+                    resourceName: 'service-role/AWSLambdaBasicExecutionRole',
+                })
+            ]
+        });
+
+        // handler allows to put bucket notification on s3 buckets.
+        role.addToPolicy(new cdk.PolicyStatement()
+            .addAction('s3:PutBucketNotification')
+            .addResource('*'));
+
+        const resource = new cdk.Resource(this, 'Resource', {
+            type: 'AWS::Lambda::Function',
+            properties: {
+                Description: 'AWS CloudFormation handler for "Custom::S3BucketNotifications" resources (@aws-cdk/aws-s3)',
+                Code: { ZipFile: `exports.handler = ${handler.toString()};` },
+                Handler: 'index.handler',
+                Role: role.roleArn,
+                Runtime: 'nodejs8.10',
+                Timeout: 300,
+            }
+        });
+
+        this.functionArn = resource.getAtt('Arn');
+    }
+}
+
+// tslint:disable:no-console
+
+/**
+ * Lambda event handler for the custom resource. Bear in mind that we are going
+ * to .toString() this function and inline it as Lambda code.
+ *
+ * The function will issue a putBucketNotificationConfiguration request for the
+ * specified bucket.
+ */
+const handler = (event: any, context: any) => {
```

Comment:
Did I not see you talk about deprecating the `InlineJavascriptLambda` in another PR?